### PR TITLE
Add KetherNFTPublisher

### PR DESCRIPTION
**Pull request by shazow**
Ready to review, mainly src/ and test/.

Rebased on an empty branch so that there's a PR/diff to review.

Migrated into a standalone repo from: https://github.com/thousandetherhomepage/ketherhomepage/pull/137

### CODE REVIEW

File: src/KetherNFTPublisher.sol [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+ *      KetherNFT.approve(address(KetherNFTPublisher), tokenId);
+ * 2. Owner of KetherNFT sets approvals on KetherNFTPublisher instance for
+ *    approved publisher:
+ *      KetherNFTPublisher.approve(msg.sender, tokenId);
+ * 3. Approved publisher can use:
+ *      KetherNFTPublisher.publish(tokenId, ...)
+ * 4. Owner can clear approval:
+ *      KetherNFTPublisher.approve(address(0), tokenId);
+ * 5. Owner can approve the magistrate:
+ *      KetherNFTPublisher.approve(address(ketherSortition), tokenId);
```
Comment (shazow) [inline src/KetherNFTPublisher.sol:46]: This logic is a bit complicated, plz doublecheck

File: src/KetherNFTPublisher.sol [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        if (publishFeeAmount > 0 && address(publishFeeToken) != address(0)) {
+            address tokenOwner = ketherNFT.ownerOf(_idx);
+            publishFeeToken.transferFrom(_msgSender(), tokenOwner, publishFeeAmount);
+        }
+
+        IKetherNFTPublish(address(ketherNFT)).publish(_idx, _link, _image, _title, _NSFW);
+    }
+
+
+    /*** Magistrate-managed Functions: ***/
```
Comment (shazow) [inline src/KetherNFTPublisher.sol:200]: Snuck in some fresh magistrate-owned mechanisms here, will allow for nice composability with some future hypothetical ERC20 😐

File: src/KetherNFTPublisher.sol [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    }
+
+    /**
+     * @dev Returns whether `publisher` is allowed to manage `tokenId`.
+     */
+    function isApprovedToPublish(address publisher, uint256 tokenId) public view returns (bool) {
+        address owner = ketherNFT.ownerOf(tokenId); // Implicitly checks tokenId validity
+        address approved = ketherNFT.getApproved(tokenId);
+
+        // Is this contract approved to control publishers?
```
Comment (mveytsman) [inline src/KetherNFTPublisher.sol]:
```suggestion
        // This contract must be either approved to publish for the token or approved for all tokens of the owner.
```

I had some trouble parsing the `!foo && !bar` construction so updated the comment.

File: src/KetherNFTPublisher.sol [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        if (owner == publisher ||
+            approved == publisher ||
+            ketherNFT.isApprovedForAll(owner, publisher)
+        ) {
+            return true;
+        }
+
+        // Is Magistrate allowed to publish?
+        address magistrate = ketherSortition.getMagistrate();
+        if (magistrate != publisher) {
```
Comment (mveytsman) [inline src/KetherNFTPublisher.sol:158]: Does this mean we can't use this contract to assign publishing to other addresses that aren't the magistrate?

**Review by mveytsman — COMMENTED** [review]
Did a preliminary pass through

My main point of misunderstanding is: 

```
 * @notice KetherNFTPublisher allows an owner of a KetherNFT to delegate
 * publishing permission to another address. Only one address can be approved
 * at a time.
```

I'm surprised that `isApprovedToPublish` is checking whether the publisher is a magistrate given the above.

There's some complicated logic here, I want to do another reading / want to find time to sync?

File: src/KetherNFTPublisher.sol (hunk 4 above)
Comment (shazow) [reply src/KetherNFTPublisher.sol:158]: This is just if all other conditions fail, it checks if the candidate is a magistrate. The non-magistrate stuff is above (and short circuits true).

File: src/KetherNFTPublisher.sol [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+            if (_tokenToPublishTimestamp[_idx] + publishTimeout > block.timestamp) {
+                revert(Errors.MustTimeout);
+            }
+            _tokenToPublishTimestamp[_idx] = block.timestamp;
+        }
+
+        // Check fee
+        if (publishFeeAmount > 0 && address(publishFeeToken) != address(0)) {
+            address tokenOwner = ketherNFT.ownerOf(_idx);
+            publishFeeToken.transferFrom(_msgSender(), tokenOwner, publishFeeAmount);
```
Comment (mveytsman) [inline src/KetherNFTPublisher.sol:193]: There might be a re-entry here.

**Review by mveytsman — APPROVED** [review]

**Pull request merged**